add single linked list implementation

julien-truffaut opened the pull request:
So the main drawbacks of `IList` at the moment is `take` and `map`, `scala.List` manage to avoid the `reverse` call. Any idea how can we improve it?

I guess one of the thing we want to debate is variance. I generally prefer invariance for collection as it makes methods simpler and we can use `widen` if we need to generalise a type. 

Here are the results of the benchmarks with jmh, units are in micro seconds. You can reproduce it with the following commands:
```
sbt
project bench
run -i 10 -wi 3 -f3 .*IList.* -rf=csv
```

| Benchmark  | size | scala.List (μs / ops) | cats.IList (μs / ops) | IList / std (%) | 
|------------:|-------------:|------------:|------------:|-------------:| 
| append | 10   | 0.13   | 0.097 | 74.62  | 
| append | 100  | 0.92   | 0.932 | 101.30 | 
| append | 1000 | 10.905 | 9.482 | 86.95  | 
| concat     | 10          | 0.207335   | 0.073252   | 35.33       | 
| concat     | 100         | 2.031151   | 0.766896   | 37.76       | 
| concat     | 1000        | 20.600249  | 8.328704   | 40.43       | 
| apply      | 10          | 0.317726   | 0.136024   | 42.81       | 
| apply      | 100         | 2.170016   | 2.050419   | 94.49       | 
| apply      | 1000        | 24.49457   | 24.7542    | 101.06      | 
| drop       | 10          | 0.014445   | 0.021218   | 146.89      | 
| drop       | 100         | 0.225092   | 0.189074   | 84.00       | 
| drop       | 1000        | 0.481145   | 0.637766   | 132.55      | 
| eq         | 10          | 0.008883   | 0.010637   | 119.75      | 
| eq         | 100         | 0.008586   | 0.010445   | 121.65      | 
| eq         | 1000        | 0.009039   | 0.010198   | 112.82      | 
| filter     | 10          | 0.292535   | 0.158818   | 54.29       | 
| filter     | 100         | 2.271045   | 1.68838    | 74.34       | 
| filter     | 1000        | 27.298933  | 16.222484  | 59.43       | 
| flatMap    | 10          | 3.11652    | 2.376278   | 76.25       | 
| flatMap    | 100         | 30.735307  | 26.094192  | 84.90       | 
| flatMap    | 1000        | 313.148512 | 227.498057 | 72.65       | 
| headOption |             | 0.010282   | 0.011587   | 112.69      | 
| lastOption | 10          | 0.02527    | 0.032417   | 128.28      | 
| lastOption | 100         | 0.241623   | 0.248764   | 102.96      | 
| lastOption | 1000        | 3.139739   | 2.848177   | 90.71       | 
| lookup     | 10          | 2.162461   | 0.01896    | 0.88        | 
| lookup     | 100         | 2.763951   | 0.215389   | 7.79        | 
| lookup     | 1000        | 1.740375   | 1.676831   | 96.35       | 
| map        | 10          | 0.215654   | 0.235267   | 109.09      | 
| map        | 100         | 1.853072   | 3.383963   | 182.61      | 
| map        | 1000        | 20.081988  | 38.771078  | 193.06      | 
| reverse    | 10          | 0.109005   | 0.109259   | 100.23      | 
| reverse    | 100         | 0.898109   | 0.994553   | 110.74      | 
| reverse    | 1000        | 10.897644  | 10.557676  | 96.88       | 
| size       | 10          | 0.021944   | 0.018177   | 82.83       | 
| size       | 100         | 0.272084   | 0.242095   | 88.98       | 
| size       | 1000        | 2.949996   | 2.831864   | 96.00       | 
| take       | 10          | 0.129515   | 0.218781   | 168.92      | 
| take       | 100         | 1.23348    | 2.308862   | 187.18      | 
| take       | 1000        | 6.714902   | 13.613678  | 202.74      | 
| prepend   | N/A       | 0.006   |  0.006   |       |

non commented:
Is your `append` operation really `prepend` (e.g. `a :: b`)?

non commented:
Also, looks like is a real build failure around the change from `Lazy#force` to `Lazy#value`.

tpolecat commented:
Yeah they do it by cheating. The implementation can modify the tail. I'm not willing to go this far.

```scala
final case class ::[B](override val head: B, private[scala] var tl: List[B]) extends List[B] { ...
```

tpolecat commented:
@julien-truffaut and I'm all in favor of invariance here.

julien-truffaut commented:
@non I renamed `append` to `concat` and cons to `prepend` but I am not entierly it is what you suggested. 

@tpolecat I agree it is a trick I am also not willing to make.

julien-truffaut commented:
btw, I am not extremely fond of `IList` name, I thought about `LinkedList` but thought it was too long. Any opinion?

gseitz commented on data/src/main/scala/cats/data/IList.scala:
```diff
@@ -0,0 +1,276 @@
+
+  /** create an [[IList]] with a single element */
+  def singleton[A](a: A): IList[A] =
+    ICons(a, INil())
```
Could use `empty[A]` here, since it's used in other methods like `filter`, `flatMap`, etc.

julien-truffaut replied:
good point

non commented on data/src/main/scala/cats/data/IList.scala:
```diff
@@ -0,0 +1,276 @@
+sealed abstract class IList[A] extends Product with Serializable {
+  import cats.data.IList._
+
+  /** add an [[IList]] to the front */
```
Most readers expect to read `as ::: bs` as "put the *as* first, then the *bs*". I think we should have `as concat bs` be read similarly (prepend the `as` on the `bs).

Right now `as concat bs` reads as "put the *bs* first then the *as*".

non commented on data/src/main/scala/cats/data/IList.scala:
```diff
@@ -0,0 +1,276 @@
+
+  /** provide a textual representation of an [[IList]] */
+  final def show(implicit A: Show[A]): String =
+    "[" + toNel.fold("")(_.reduceLeft(A.show)(_ + "," + A.show(_))) + "]"
```
Would you consider using a more efficient `StringBuilder`-based method here?

julien-truffaut replied:
sure, I let that for later as I considered `show` less performance critical

non commented on data/src/main/scala/cats/data/IList.scala:
```diff
@@ -0,0 +1,276 @@
+
+  /** transform an [[IList]] into a [[scala.List]] */
+  final def toList: List[A] =
+    reverse.foldLeft(List.empty[A])((acc, a) => a :: acc)
```
You could speed this up by using `foldLeft(ListBuffer.empty[A])(...)` instead of having to reverse the `IList`.

non commented:
Would anyone else prefer to have the following type alias:

```scala
type NEL[A] = ICons[A]
```

I've always hated the `(A, List[A])` representation for non-empty lists. I find using the cons type for this a lot more elegant. What do people think?

mpilquist commented:
@non I like `type NEL[A] = ICons[A]`

non commented:
@mpilquist Cool. I realized belatedly that maybe I shouldn't have made this comment on @julien-truffaut's PR -- I don't think it's his job to implement that.

tpolecat commented:
@non I'm nervous about `type NEL[A] = ICons[A]` for a few reasons.

- We're treating a data constructor as if it were a real type, which is something we try not to do in general; usually if you end up with something of type `Some[Woozle]` you're misunderstanding things or have experienced an inference failure. So I'm not sure this is a desirable precedent.
- `NEL` and `IList` have a largely disjoint set of typeclass instances, and derive different instances for instantiated types. For instance, `NEL` is a `Comonad` and `Foldable1`, and `IList` is a `MonoidK` or whatever we're calling it. `IList[A]` is `Monoid` for all `A` but `NEL` isn't. So I fear this will lead to suffering at some point.
- This encoding would require all the methods on `ICons` like `map` and so on to return an `ICons`, otherwise your `NEL` will keep turning into an `IList`. So it may lead to a confusing implementation. It also means we can't implement `filter` at all, as far as I can tell, since you would be able to call it on an `NEL` ... I guess you get an `IList` back? Seems very weird to me.

My fears may not be justified (I haven't had any coffee yet so I may revise and extend my remarks in a little bit) but it kind of smells like trouble.

non commented:
Yeah I guess I agree with this. I proposed another encoding on Gitter which might work instead.

tpolecat commented:
@non proposed the following alternate encoding:

```scala
sealed trait IList[A]
sealed trait NEL[A]
case class Cons[A](head: A, tail: IList[A]) extends IList[A] with NEL[A]
case class INil[A]() extends IList[A]
```

It seems odd to me that the `IList` and `NEL` share a data constructor, but otherwise I think this encoding addresses my concerns above. `IList` and `NEL` will have many identical methods (`Listy[A]` supertrait?), and many that differ only in return type, but otherwise it seems to work. Exhaustiveness checking works correctly based on the static type of the `Cons`, as expected.

I'm not quite convinced that this is any better than `OneAnd[List,A]` or `Cofree[Option,A]` which give you the `Comonad` and some other stuff for free ... or even a custom type ... but it seems worthwhile to mess with it and see how it turns out.

paulp commented:
@tpolecat Not only do they cheat, [they get caught](https://issues.scala-lang.org/browse/SI-7838).

And let's not forget they're ummutable:
```
scala> val xs = scala.collection.mutable.ListBuffer(1, 2, 3)
xs: scala.collection.mutable.ListBuffer[Int] = ListBuffer(1, 2, 3)

scala> val ys = xs.toIterable match { case xs: List[Int] => xs }
ys: List[Int] = List(1, 2, 3)

scala> xs append (4, 5, 6)

scala> ys
res0: List[Int] = List(1, 2, 3, 4, 5, 6)

scala>
```

refried commented:
This var revelation makes me sad.

julien-truffaut commented:
I think I addressed the comments except for `show`, we should investigate @non proposition regarding `Nel` but I suppose it is better in a separate PR.

non commented:
Great. I'm happy with this for now, it provides a good base to work on.

I'm also pleased that it provides a good guide for benchmarks that other data types can use. Thanks!

:+1:

ceedubs commented:
As briefly discussed on Gitter, I think we should have a discussion about if/how/how-much we are going to provide alternatives to widely-used structures in the std lib. I'm not sure pull request comments are the best place. Maybe in the next Cats Google Hangout session?

refried commented:
I think it's good to capture the discussion here, even if it means someone acts as Hangouts Secretary and summarizes the emerging viewpoints, on the ticket?

julien-truffaut commented:
I am fine with both solutions (using this PR or wait for next hangout)

tpolecat commented:
Just mark me down as happy to throw stdlib to the sharks. `-Yno-imports` \o/

Anyway :+1: on this. Love the benchmarking.

julien-truffaut commented:
@tpolecat I would also like that we propose a serious alternative to std collection

paulp commented:
@tpolecat @julien-truffaut I hope you will take a look at what I've already done in this direction. 

[Layer 1](https://github.com/paulp/psp-std/tree/master/dmz/src/main/scala)
[Layer 2](https://github.com/paulp/psp-std/tree/master/std/src/main/scala), most of which is inapplicable here but some of which goes with Layer 1.

The separation is primarily to work around the nightmarish non-deterministic compilation issues which arise during package object composition when one is operating under -Yno-predef -Yno-imports.

I am amenable to extracting an agreeable subset of this to provide us with a common basis, if such a subset can be identified.

mpilquist commented:
I'm all for having an alternative collection library, though I'm not convinced that such a library belongs in cats. I'd imagine that the rate of change in a collection library will be higher than in cats. I'm fairly neutral on `IList` and `Or` being in cats though. Perhaps they are so fundamental that they should exist in cats-core while other general purpose collections get their own library.

non commented:
I agree with @mpilquist that I'd prefer for Cats not to provide a full replacement for the collections library. There are a few reasons for this:

1. I want Cats to benefit users who stick to the standard collections. Thus, I don't want to write our APIs in terms of our own custom collections (unless absolutely necessary).
2. I don't see major downsides to doing this work in a separate library.
3. I feel like the whole idea of a *collections library* may in fact be too large. I could imagine a bottom-up project of implementing many concrete data structures that have the characteristics we want, and then figuring out how to abstract over them (with type classes, class hierarchy, etc).

Anyway, here are the sorts of data structures I'd prefer to include:

1. Immutable data structures that are *missing* from the standard library and widely useful (e.g. `Dequeue`, immutable arrays)
2. Data structures which are needed by other type classes or APIs (e.g. `Kleisli` or `Lazy`)

Personally, while `Option` has many warts, I don't think it's worth including an alternate option type in Cats. I think it is possible to write safe, idiomatic, functional code using `Option`, and it is already widely-used. The cost of conversion between a reimplementation and `scala.Option` would be too high.

By contrast, I think reimplementing something like `scala.Either` makes sense. The existing type is not biased (and is designed/documented such that adding an implicit bias would be confusing). Also it isn't widely-used in public APIs, meaning there is less chance of massive amounts of conversion code back and forth between them.

I'm open to including list (especially if we have a good story for interop with non-empty lists too) but I'm a bit more torn.

Does this make sense? It's just my 2¢ for Cats -- I would love to see libraries that do a top-to-bottom reimplementation of all of it and I think it's valuable work.

adelbertc commented:
> I want Cats to benefit users who stick to the standard collections. Thus, I don't want to write our APIs in terms of our own custom collections (unless absolutely necessary). In fact, I'd like to limit the extend to which we mention concrete collections in our APIs at all.

+1, agreed with that. While for my personal projects I often reach for `scalaz` collections, at $JOB I use standard collections so I sympathize with this point.

I think putting it in a separate library as suggested may be a good idea in the effort for modularization, but I'd still like to see things like an alternative `scala.Either` in the core cats lib.. where to draw the line gets tricky, but the proposal of

> Immutable data structures that are missing from the standard library and widely useful (e.g. Dequeue, immutable arrays, etc.) Data structures which are needed by other type classes or APIs (e.g. Kleisli or Lazy)

Looks good to me. For me I'd put the list in `cats-data` or whatnot.

milessabin commented:
@non +ω.

julien-truffaut commented:
I also agree that `cats` is not the right place for another implementation of `List` or the rest of scala collection. I quite like @non criteria, especially the second point, regarding which data structure should we have in `cats-data`.
